File: csharp/ToolGood.Algorithm/Internals/Functions/MathSum2/Function_EXPONDIST.cs
using System;
using System.Collections.Generic;
using System.Text;
using ToolGood.Algorithm.Enums;
using ToolGood.Algorithm.Internals;
using ToolGood.Algorithm.MathNet.Numerics;

namespace ToolGood.Algorithm.Internals.Functions.MathSum2
{
	internal sealed class Function_EXPONDIST : Function_3
	{
		public Function_EXPONDIST(FunctionBase[] funcs) : base(funcs)
		{
		}



		public override string Name => "ExpOnDist";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetNumber_1(engine, tempParameter);
			if(args1.IsErrorOrNone) return args1;

			var args2 = GetNumber_2(engine, tempParameter);
			if(args2.IsErrorOrNone) return args2;

			var args3 = GetBoolean_3(engine, tempParameter);
			if(args3.IsErrorOrNone) return args3;

			var n1 = args1.NumberValue;
			var rate = args2.NumberValue;
			if(n1 < 0.0m) {
				return ParameterError(1);
			}
			if(rate <= 0.0m) {
				return ParameterError(2);
			}
			return Operand.Create(ExponDist(n1, rate, args3.BooleanValue));
		}

		public decimal ExponDist(decimal x, decimal rate, bool state)
		{
			if(state) {
				return CDF(rate, x);
			}
			return PDF(rate, x);
		}
		public decimal PDF(decimal rate, decimal x)
		{
			return x < 0.0m ? 0.0m : rate * MathEx.Exp(-rate * x);
		}

		public decimal CDF(decimal rate, decimal x)
		{
			return x < 0.0m ? 0.0m : 1.0m - MathEx.Exp(-rate * x);
		}

		public override OperandType GetResultType()
		{
			return OperandType.NUMBER;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
			func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
			func3.GetParameterTypes(noneEngine, result, OperandType.BOOLEAN);
		}
	}

}

File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum2/Function_EXPONDIST.java
package toolgood.algorithm.internals.functions.mathsum2;

import java.math.BigDecimal;
import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.mathNet.ExcelFunctions;

public final class Function_EXPONDIST extends Function_3 {
    public Function_EXPONDIST(FunctionBase func1, FunctionBase func2, FunctionBase func3) {
        super(func1, func2, func3);
    }

    public Function_EXPONDIST(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "ExponDist";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        Operand args2 = GetNumber_2(engine, tempParameter);
        if (args2.IsErrorOrNone()) {
            return args2;
        }
        Operand args3 = GetBoolean_3(engine, tempParameter);
        if (args3.IsErrorOrNone()) {
            return args3;
        }

        double x = args1.DoubleValue();
        double lambda = args2.DoubleValue();
        boolean cumulative = args3.BooleanValue();

        if (x < 0.0 || lambda <= 0.0) {
            return Operand.Error("Function '{0}' parameter is error!", "ExponDist");
        }
        return Operand.Create(ExcelFunctions.ExponDist(x, lambda, cumulative));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func3.GetParameterTypes(noneEngine, result, OperandType.BOOLEAN);
    }
}
